### ISSUE TITLE
Gatsby 4 - EPIPE error on Netlify

### ISSUE THREAD

**Issue by SamHoltby** [gatsbyjs/gatsby#33738]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Just upgraded to Gatsby 4 and everything is working fine locally. I can npm develop and build without issues. However, as soon as I push to Netlify the below error is returned near the end of building.

```    4:07:59 PM: success run queries in workers - 158.327s - 621/621 3.92/s
    4:16:57 PM: error UNHANDLED EXCEPTION write EPIPE
    4:16:57 PM: 
    4:16:57 PM: 
    4:16:57 PM:   Error: write EPIPE
    4:16:57 PM:   
    4:16:57 PM:   - child_process:846 ChildProcess.target._send
    4:16:57 PM:     node:internal/child_process:846:20
    4:16:57 PM:   
    4:16:57 PM:   - child_process:719 ChildProcess.target.send
    4:16:57 PM:     node:internal/child_process:719:19
    4:16:57 PM:   
    4:16:57 PM:   - index.js:290 WorkerPool.sendMessage
    4:16:57 PM:     [repo]/[gatsby-worker]/dist/index.js:290:19
    4:16:57 PM:   
    4:16:57 PM:   - worker-messaging.ts:22 
    4:16:57 PM:     [repo]/[gatsby]/src/utils/jobs/worker-messaging.ts:22:22
    4:16:57 PM:   
    4:16:57 PM: 
    4:16:57 PM: not finished Merge worker state - 0.168s
```

The Netlify site is deploying on node v16.13.0 (npm v8.1.0) - which is the same as my local environment(Windows 10).

Netlify support believe it's an issue with Gatsby, which surely can't be the case if it's working locally?

### Reproduction Link

Unable to provide

### Steps to Reproduce

1. Install the following packages after cloning https://github.com/gatsbyjs/gatsby-starter-wordpress-blog :
gatsby: ^4.0.2 => 4.0.2
    gatsby-background-image: ^1.5.3 => 1.5.3
    gatsby-plugin-feed: ^4.0.0 => 4.0.0
    gatsby-plugin-google-tagmanager: ^4.0.0 => 4.0.0
    gatsby-plugin-image: ^2.0.0 => 2.0.0
    gatsby-plugin-manifest: ^4.0.0 => 4.0.0
    gatsby-plugin-nprogress: ^4.0.0 => 4.0.0
    gatsby-plugin-offline: ^5.0.0 => 5.0.0
    gatsby-plugin-react-helmet: ^5.0.0 => 5.0.0
    gatsby-plugin-react-svg: ^3.0.1 => 3.1.0
    gatsby-plugin-remove-fingerprints: ^0.0.2 => 0.0.2
    gatsby-plugin-sass: ^5.0.0 => 5.0.0
    gatsby-plugin-sharp: ^4.0.1 => 4.0.1
    gatsby-plugin-sitemap: ^5.0.0 => 5.0.0
    gatsby-plugin-styled-components: ^5.0.0 => 5.0.0
    gatsby-plugin-svgr-loader: ^0.1.0 => 0.1.0
    gatsby-plugin-web-font-loader: ^1.0.4 => 1.0.4
    gatsby-source-filesystem: ^4.0.0 => 4.0.0
    gatsby-source-wordpress: ^6.0.0 => 6.0.0
    gatsby-transformer-sharp: ^4.0.0 => 4.0.0

2. Setup WordPress installation and configure with the following gatsby-config:
```
 {
    resolve: `gatsby-source-wordpress`,
    options: {
      url: process.env.GRAPHQL_API,
      debug: {
        graphql: {
          writeQueriesToDisk: true,
        },
      },
      production: {
        allow404Images: true,
      },
      auth:
        process.env.NODE_ENV === "production"
          ? {
              htaccess: {
                username: process.env.HTACCESS_USERNAME,
                password: process.env.HTACCESS_PASSWORD,
              },
            }
          : {},
      type: {
        Post: {
          limit:
            process.env.NODE_ENV === `development`
              ? 30
              : process.env.POST_LIMIT || 999999,
        },
        Comment: {
          exclude: true,
        },
        PostFormat: {
          exclude: true,
        },
      },
    },
  },
```

3. Deploy to Netlify

### Expected Result

Site builds successfully

### Actual Result

Site build fails each time with `UNHANDLED EXCEPTION write EPIPE` error.

### Environment

```shell
My local environment: 

 System:
    OS: Windows 10 10.0.19042
    CPU: (16) x64 Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz
  Binaries:
    Node: 16.13.0 - C:\Program Files\nodejs\node.EXE
    npm: 8.1.0 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.10.0
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (95.0.1020.30)
  npmPackages:
    gatsby: ^4.0.2 => 4.0.2
    gatsby-background-image: ^1.5.3 => 1.5.3
    gatsby-plugin-feed: ^4.0.0 => 4.0.0
    gatsby-plugin-google-tagmanager: ^4.0.0 => 4.0.0
    gatsby-plugin-image: ^2.0.0 => 2.0.0
    gatsby-plugin-manifest: ^4.0.0 => 4.0.0
    gatsby-plugin-nprogress: ^4.0.0 => 4.0.0
    gatsby-plugin-offline: ^5.0.0 => 5.0.0
    gatsby-plugin-react-helmet: ^5.0.0 => 5.0.0
    gatsby-plugin-react-svg: ^3.0.1 => 3.1.0
    gatsby-plugin-remove-fingerprints: ^0.0.2 => 0.0.2
    gatsby-plugin-sass: ^5.0.0 => 5.0.0
    gatsby-plugin-sharp: ^4.0.1 => 4.0.1
    gatsby-plugin-sitemap: ^5.0.0 => 5.0.0
    gatsby-plugin-styled-components: ^5.0.0 => 5.0.0
    gatsby-plugin-svgr-loader: ^0.1.0 => 0.1.0
    gatsby-plugin-web-font-loader: ^1.0.4 => 1.0.4
    gatsby-source-filesystem: ^4.0.0 => 4.0.0
    gatsby-source-wordpress: ^6.0.0 => 6.0.0
    gatsby-transformer-sharp: ^4.0.0 => 4.0.0
  npmGlobalPackages:
    gatsby-cli: 4.0.0

Netlify site:
    node v16.13.0 (npm v8.1.0) 
    Using ruby version 2.7.1
    Using PHP version 8.0
```

**Comment by benlavalley** [comment]
Running into this as well with my Netlify site after updating to Gatsby v4 (node 14.18.1 installed locally and  specified in .nvmrc)

**Comment by LekoArts** [comment]
Hi!

Could you please [provide a minimal reproduction][reproduction] so that we can use that repo to run it on Netlify itself? Can you also link the issue/discussion you opened with Netlify support (if public)? Thanks!

[reproduction]: https://gatsby.dev/reproduction

**Comment by fool** [comment]
I'm from Netlify Support and came to see if y'all had managed to track this down.  Sam's question at least isn't public (it's in our helpdesk), but I have linked this issue to it on our side.

What I can see without betraying Sam's confidence is that it happens right after this phase of the build completes:

`success run queries in workers - x.ys 621/621 4.08/s`.  I guess Sam may have to tell us what the next step is, since that (9 minutes later) seems to be the step that fails.

@benlavalley what is the behavior of your failure?  Is it also after a long delay in a step other than run queries?

**Comment by SamHoltby** [comment]
@LekoArts will get something up ASAP, unable to at the moment (traveling).

@fool Thanks for the comment. As I mentioned in the help desk ticket, this doesn't happen on my local environment, making it tricky to debug. The main difference is Gatsby and their plugin versions.

@benlavalley Had any luck debugging this? Let us know if you find a solution 🙂

**Comment by benlavalley** [comment]
@fool  Yes it does happen after a period of processing time.

`
1:43:47 PM: success run queries in workers - 139.530s - 340/340 2.44/s
1:47:45 PM: error UNHANDLED EXCEPTION write EPIPE
1:47:45 PM: 
1:47:45 PM: 
1:47:45 PM:   Error: write EPIPE
1:47:45 PM:   
1:47:45 PM:   - child_process.js:841 ChildProcess.target._send
1:47:45 PM:     internal/child_process.js:841:20
1:47:45 PM:   
1:47:45 PM:   - child_process.js:712 ChildProcess.target.send
1:47:45 PM:     internal/child_process.js:712:19
1:47:45 PM:   
1:47:45 PM:   - index.js:290 WorkerPool.sendMessage
1:47:45 PM:     [repo]/[gatsby-worker]/dist/index.js:290:19
1:47:45 PM:   
1:47:45 PM:   - worker-messaging.ts:22 
1:47:45 PM:     [repo]/[gatsby]/src/utils/jobs/worker-messaging.ts:22:22
1:47:45 PM:   
1:47:45 PM: 
1:47:45 PM: not finished Merge worker state - 0.152s
1:47:45 PM: ERROR: "build:app" exited with 1.`

For anyone from Netlify investigating, feel free to dig into anything with my account -- my netlify site name is amazing-wiles-1918a6 and the published URL is https://floridabulldogshockey.com. My first React / Gatsby site. 

As I have time I'll work on trying to get a more simple reproduction put together.

**Comment by wardpeet** [comment]
Could you check peak RAM usage? My guess is that the worker crashed due to OOM. EPIPE usually means that the communication channel got destroyed for any reason.

**Comment by benlavalley** [comment]
> Could you check peak RAM usage? My guess is that the worker crashed due to OOM. EPIPE usually means that the communication channel got destroyed for any reason.

This seems like a culprit.

I just drastically reduced the static file content footprint of my site to test, and it deployed successfully. 

When I added my content back, the deploy fails again. 

@fool Can you bump my memory limit up? :)

**Comment by benlavalley** [comment]
Upon further testing and following some suggestions from the Netlify support team, I was able to run my [own Netlify local build processes](https://github.com/netlify/build-image) and attempt to reproduce the Netlify production build error I was seeing.

Mission success -- the issue was reproduced and I seemed to get the same write EPIPE error.

Upon bumping my Docker memory limit up from 3GB to 10GB, my build succeeds.

I'd guess it is only natural that Gatsby v4 uses more memory with it's SSR enhancements over Gatsby v3.

I put together some really clever routines to parse file-based content with Gatsby and offer up what would otherwise seem to be server-side generated content (I have drop-down filters, searching, and paging for online shop and sports team roster content -- all statically generated) but it seems like this creating generating too much data or too many pages for Netlify to support with the build process 3gb memory limit :( 

More detail in the Netlify support forum here: https://answers.netlify.com/t/gatsby-v4-works-locally-but-timed-out-on-netlify/46339/9

Netlify states they can't budge on their 3GB memory limit, so I have to figure out how I'm going to move forward now -- do I rearchitect my site to get rid of a ton of clever work that went into building an interactive site that was backed by the performance of statically generated content, or do I get rid of it all and dive into Gatsby's SSR support and then *hope* that my builds stay under Netlify's 3GB memory limit...

**Comment by SamHoltby** [comment]
@benlavalley Thanks for doing all the legwork! 

It's interesting that Netlify won't budge on the 3GB memory limit. On their pricing page, via the [High-Performance Build](https://www.netlify.com/pricing/#business-products-high-performance-build) section, they state Up to 6GB of memory...

I've asked Netlify to increase my memory. If it's possible I'll update this issue with the results.

**Comment by benlavalley** [comment]
> @benlavalley Thanks for doing all the legwork!
> 
> It's interesting that Netlify won't budge on the 3GB memory limit. On their pricing page, via the [High-Performance Build](https://www.netlify.com/pricing/#business-products-high-performance-build) section, they state Up to 6GB of memory...
> 
> I've asked Netlify to increase my memory. If it's possible I'll update this issue with the results.

Maybe I'm misreading it, but it looks like I'm supposed to have "up to" 6GB on that pricing page? I did try updating my Netlify build image from Xenial to Focal, crossing my fingers and thinking maybe I was on some sort of older system, but no luck.

Based on my reading here, @SamHoltby we may have luck attempting builds late at night when Netlify's systems are less busy - they guarantee 3GB, but it seems we may get "up to" 6GB if resources allow.

**Comment by benlavalley** [comment]
I made some changes to my app, removed content and slowly re-added it, and I seem to be able to deploy now, though I don't think it will work consistently due to the variable nature of memory allocation in Netlify.

For anyone coming across this, changes I made to my app to try and improve efficiency/lower memory usage:

Gatsby-specific:
- Migrate from gatsby-image to gatsby-plugin-image
- Removed a chunk of unecessary images from processing
- Tried adding GATSBY_CONCURRENT_DOWNLOAD environment variable and set to 15 (default 200)
- Set GATSBY_CPU_COUNT to 1 for production deploys (attempts to reduce parallelization/peak memory usage)

Netlify-specific:
- Added gatsby-plugin-remove-fingerprints package in my netlify config
- Migrated from Xenial to Focal build image
- Added @netlify/plugin-gatsby

I have a feeling that if my site's cache is cleared I might see deployment failures again. Crossing fingers, and I hope some of this might help Sam and any others that run into this problem.

**Comment by SamHoltby** [comment]
@benlavalley sorry for the radio silence. I ended up moving to Gatsby Cloud for the builds and Netlify for the hosting. Didn't need to change anything, plus Gatsby builds 70% faster.

**Comment by benlavalley** [comment]
> @benlavalley sorry for the radio silence. I ended up moving to Gatsby Cloud for the builds and Netlify for the hosting. Didn't need to change anything, plus Gatsby builds 70% faster.

Interesting! I actually explored doing the same thing. I've liked Netlify. Gatsby's equivalent tier to what Netlify provides now seems to state it has 1GB of memory which is less than Netlify's 3GB. I guess I can test out a deploy at least and see how it fares - I'm back to getting the EPIPE errors with Netlify now because of the variable memory limits.

**Comment by MJDCKP** [comment]
Hi guys, do we have proper solution for this?

**Comment by TiagoJeronimo** [comment]
Hello, I've also started having the same problem recently, is there any proper solution?

**Comment by benlavalley** [comment]
@MJDCKP  @TiagoJeronimo If you are seeing recent errors as of this month, it is likely related to something introduced in recent versions of Gatsby (they attempted to reduce memory usage but it looks like something else started fouling up).

I locked my Gatsby version to =v4.1.3 in my packag.json and I can actually deploy my site now, though I still have to reduce the overall amount of static content my site includes.

**Comment by wahidshafique** [comment]
> * gatsby-plugin-remove-fingerprints

Yes I got the same issue after upgrading to the latest 4.9

**Comment by davidbbe** [comment]
I was seeing this same build error on my Netlify hosted site when I upgraded to Gatsby 4.9. I saw a couple people [mention](https://answers.netlify.com/t/gatsby-v4-works-locally-but-timed-out-on-netlify/46339/18) lowering `GATSBY_CONCURRENT_DOWNLOAD`. I added this to my Netlify environment variables with a value of 16 and my builds complete without errors and with no other changes.

**Comment by natapg** [comment]
> I was seeing this same build error on my Netlify hosted site when I upgraded to Gatsby 4.9. I saw a couple people [mention](https://answers.netlify.com/t/gatsby-v4-works-locally-but-timed-out-on-netlify/46339/18) lowering `GATSBY_CONCURRENT_DOWNLOAD`. I added this to my Netlify environment variables with a value of 16 and my builds complete without errors and with no other changes.

In my case, this solution did not help, nor was it solved by optimizing the images or disabling WEBP. With the latest 4.9 it gives me the error in 90% of the deployments

```
2:18:10 PM: success run queries in workers - 99.219s - 221/221 2.23/s
2:21:09 PM: success Running gatsby-plugin-sharp.IMAGE_PROCESSING jobs - 490.668s - 308/308 0.63/s
2:21:09 PM: error UNHANDLED EXCEPTION write EPIPE
2:21:09 PM: 
2:21:09 PM: 
2:21:09 PM:   Error: write EPIPE
2:21:09 PM:   
2:21:09 PM:   - child_process.js:841 ChildProcess.target._send
2:21:09 PM:     internal/child_process.js:841:20
2:21:09 PM:   
2:21:09 PM:   - child_process.js:712 ChildProcess.target.send
2:21:09 PM:     internal/child_process.js:712:19
2:21:09 PM:   
2:21:09 PM:   - index.js:291 WorkerPool.sendMessage
2:21:09 PM:     [repo]/[gatsby-worker]/dist/index.js:291:19
2:21:09 PM:   
2:21:09 PM:   - worker-messaging.ts:22 
2:21:09 PM:     [repo]/[gatsby]/src/utils/jobs/worker-messaging.ts:22:22
2:21:09 PM:   
2:21:09 PM: 
2:21:09 PM: not finished Merge worker state - 0.093s
2:21:09 PM: error Command failed with exit code 1. (https://ntl.fyi/exit-code-1)
```

**Comment by natapg** [comment]
> 
> What's the large image you're processing (byte size and resolution)?
> 
> For me reducing image size and setting the following node memory configuration in the Netlify ENV variables, fixed the issue:
> 
> `NODE_OPTIONS = --max-old-space-size= 4096`

I also tried adding that variable without luck! 
The largest image is 580 KB and 2240 x 1260

**Comment by jessmartin** [comment]
> I locked my Gatsby version to =v4.1.3 in my packag.json and I can actually deploy my site now, though I still have to reduce the overall amount of static content my site includes.

Just wanted to comment that I was also running into this issue, and @benlavalley's above advice fixed it for me. 👏👏👏

**Comment by grubersjoe** [comment]
I have two Gatsby pages. Both their build pipeline on Netlify stopped working after Upgrading Gatsby (building locally still works). I've tried pretty much all the suggestions in this threads, nothing helps. I think lots of people might be affected by this issue given how popular Netlify is for continuous deployment. So what can we do? It used to work fine, moving to Gatsby Cloud is not an option for me.

**Comment by t2ca** [comment]
> I have two Gatsby pages. Both their build pipeline on Netlify stopped working after upgrading Gatsby (building locally still works). I've tried pretty much all the suggestions in this thread, nothing helps. I think lots of people might be affected by this issue given how popular Netlify is for continuous deployment. So in my eyes this is a pretty serious bug. What can we do? It used to work fine. The proposed "solution" to move to Gatsby Cloud is not an option for me.
> 
> My current workaround is to build the page locally and to execute `netlify deploy --prod` then.

Have you tried to pin gatsby to `v4.7.2`? This is a solution that has worked for me.

We are also having a discussion about this here https://github.com/gatsbyjs/gatsby/issues/35055

I should also mention that the Gatsby team is looking into this issue.

**Comment by wutsqo** [comment]
I just wanna say that I faced the same issue on Gatsby v4.10

> Have you tried to pin gatsby to v4.7.2? This is a solution that has worked for me.

This solution worked for me. Thanks

**Comment by ChangerHe** [comment]
> 

This worked for me too ! (from v4.10.2 to v4.7.2)
And a small tips:  v4.7.2 not support built in gatsby-config.ts, use gatsby-config.js instead.

**Comment by ascorbic** [comment]
I'm going to repopen this, as it's clearly not fixed. I'm going to look at this more today. There's a lot more discussion on https://github.com/gatsbyjs/gatsby/issues/35055

**Comment by leils** [comment]
+1 rolling back to 4.7.2 appears to resolve this for me.

**Comment by andrewgolovanov** [comment]
in my case, version `4.7.0` helped solve the problem, on version `4.7.2` the error persisted

**Comment by danielmoessner** [comment]
I have the same problem on Gatsby Version 4.11.
Why is this issue closed?

**Comment by RazvanRauta** [comment]
Same problem: `"gatsby": "^4.12.1",`
![image](https://user-images.githubusercontent.com/40596596/163672103-dd8298bf-9a7f-47f9-8f55-b9131f72971e.png)

**Comment by samgermain** [comment]
> @benlavalley sorry for the radio silence. I ended up moving to Gatsby Cloud for the builds and Netlify for the hosting. Didn't need to change anything, plus Gatsby builds 70% faster.

Whats the benefit of hosting on Netlify instead of just hosting on Gatsby Cloud?

-----------

Also this issue shouldn't be closed, this solution is more of a hack than a solution

**Comment by DavidSabine** [comment]
It appears incontrovertible this error relates to memory usage. And many claim that freezing Gatsby at version 4.7.0 works.

Dumb question: has anyone identified the precise diff between 4.7.0 and next versions that changes the memory usage so much?

Would be great to solve this. Really frustrating.

**Comment by michaellunzer** [comment]
+1   I'm also experiencing this error and fixed it by rolling back to Gatsby v4.7.0

Looking forward to what others suggest to move forward!

**Comment by ascorbic** [comment]
Hi everyone. There is now fix out for this, so if you upgrade to `gatsby@4.13.1` it should solve the problem. This is being tracked at #35055, so please post there about whether or not this fixes it for you.
